lib/seahorn/Smt/TypeChecker.cc

```diff
@@ -12,64 +12,72 @@ namespace {
 
 //==-- main implementation goes here --==//
 class TCVR {
-  ExprVector expTypeStack;
-
-  bool isConst (Expr exp) {
-   return bind::isBoolConst(exp) || bind::isIntConst(exp);
+  ExprMap exprMap;
+
+  bool isBool(Expr exp) {
+    if (isOpX<TRUE>(exp) || isOpX<FALSE>(exp))
+      return true;
+    else if (bind::isBoolVar(exp) || bind::isBoolConst(exp))
+      return true;
+    else if (isOpX<AND>(exp) || isOpX<OR>(exp) || isOpX<XOR>(exp) ||
+             isOpX<IMPL>(exp) || isOpX<ITE>(exp) || isOpX<IFF>(exp) ||
+             isOpX<NEG>(exp))
+      return true;
+
+    return false;
   }
 
-  bool isVar (Expr exp) {
-   return bind::isBoolVar(exp) || bind::isIntVar(exp);
-  }
+  Expr inferType(Expr exp /*,TypeChecker & ty*/) {
+    if (isBool(exp))
+      return sort::boolTy(exp->efac());
```

Comment:
you also want to check that the expression is well formed -- i.e., children are of the right type and the right number of children. For example, is `NOT (A, B)` allowed? is `AND A` allowed? `AND` with no operands?

Reply:
Is it fine to return Expr() when it is not well formed or would you recommend introducing a new error type?

Reply:
I would not abuse `Expr()` too much. It is basically the same as `nullptr`. Having it all over the place will cause all sort of issues if null is not expected. For example, you would not be able to easily print type of all expressions to discover at which point type error was introduced.